Explorer update

navFooh opened the pull request.

navFooh committed:
Create useMsigGetAvailableBalance

navFooh committed:
fix MsigGetAvailableBalance

navFooh committed:
Use useMsigGetAvailableBalance

navFooh committed:
Log available balance errors

navFooh committed:
Ensure network cointype for address

navFooh committed:
ActorState.propTypes

navFooh committed:
Improve available balance display

navFooh committed:
Display Actor Name and Code

navFooh committed:
Update Lines.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.ts

navFooh committed:
Update index.ts

navFooh committed:
Update index.ts

navFooh committed:
Update index.ts

navFooh committed:
Simplify useStateReadState

navFooh committed:
Update index.ts

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Add MsigState to proposal

navFooh committed:
Update Proposal.tsx

navFooh committed:
Update index.ts

navFooh committed:
Add infoMsg for caption

navFooh committed:
fix caption props

navFooh committed:
Update index.tsx

navFooh committed:
implement actorNotFound

navFooh committed:
allow string array for isAddressSigner

navFooh committed:
Update customPropTypes.ts

navFooh committed:
Update index.tsx

navFooh committed:
Update Proposal.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Merge branch 'primary' into explorer-update

navFooh committed:
Update index.ts

navFooh committed:
Update index.ts

navFooh committed:
Update packages

navFooh committed:
Update index.test.tsx.snap

navFooh committed:
Update index.tsx

navFooh committed:
Update Lines.tsx

navFooh committed:
Update AddressLink.tsx

navFooh committed:
Update AddressLink.tsx

navFooh committed:
Update AddressLink.tsx

navFooh committed:
Create DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Use rest spread

navFooh committed:
Rename CidLink to MessageLink

navFooh committed:
Update MessageLink.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
bump filecoin-actor-utils

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Bump @glif/filecoin-actor-utils

navFooh committed:
Update Snapshots

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Merge branch 'primary' into explorer-update

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update index.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Bump @glif/filecoin-actor-utils

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Merge branch 'primary' into explorer-update

navFooh committed:
Update detail.tsx

navFooh committed:
Update index.tsx

navFooh commented on src/components/HistoryTables/MsigProposals/Proposal.tsx:
```diff
@@ -39,111 +35,96 @@ export default function ProposalDetail({
     loading: actorLoading,
+    notFound: actorNotFound,
     error: actorError
-  } = useActorQuery({
-    variables: { address: convertAddrToPrefix(msigAddress, coinType) }
-  })
```
I'm not sure why to `useActorQuery` was used here. It seemed to only grab the actor code, to confirm that it was an msig actor and then perfom `useStateReadState`, which also gets the actor code

Schwartz10 replied:
Probably because i'm dumb, but possibly because I think originally i envisioned this working in a way where a user could enter any actor address (i.e. showing the proposals on the explorer) - making the `useActorQuery` helped me get a better error about "actor not found" at the time. Good call here though

navFooh commented on src/components/HistoryTables/MsigProposals/Proposal.tsx:
```diff
@@ -39,111 +35,96 @@ export default function ProposalDetail({
-    error: stateError
-  } = useStateReadStateQuery<MsigState>({
-    variables: { address: convertAddrToPrefix(msigAddress, coinType) },
-    skip: !loadMsigState
```
Skip was not being handled by `useStateReadStateQuery`, see comment in `useStateReadState.ts`

navFooh commented on src/components/LabeledText/AddressLink.tsx:
```diff
@@ -10,45 +10,43 @@ import {
 }: AddressLinkProps) => {
   const { explorerUrl } = useEnvironment()
   const { data: gqlAddress } = useAddressQuery({
-    skip: !id || !!address || !fetchAddress,
-    variables: {
-      address: id
-    }
+    skip: !fetchAddress,
+    variables: { address }
```
Changed the logic here a bit, guessing this is what you mentioned before :)

Schwartz10 replied:
Yes

navFooh commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,204 @@
+    case Type.Object:
+      const { Name, Children } = dataType
+
+      if (Name === 'Cid') {
+        const cid = Children['/'].Value as string
+        return (
+          <DataTypeLine
+            label={`${label} (CID)`}
+            depth={depth}
+            dataType={{ Name, Type: Type.String, Value: cid }}
+          />
+        )
+      }
```
This is temporary until we really know how to unwrap the CIDs

navFooh commented on src/utils/useMsigGetAvailableBalance/index.ts:
```diff
@@ -0,0 +1,38 @@
+import { useEffect, useState } from 'react'
+import { FilecoinNumber } from '@glif/filecoin-number'
+import { validateAddressString } from '@glif/filecoin-address'
+
+import { useEnvironment } from '../../services/EnvironmentProvider'
+
+interface UseMsigGetAvailableBalanceResult {
+  availableBalance: FilecoinNumber | null
+  loading: boolean
+  error: Error | null
+}
+
+export const useMsigGetAvailableBalance = (
+  address: string
+): UseMsigGetAvailableBalanceResult => {
+  const [balance, setBalance] = useState<FilecoinNumber | null>(null)
+  const [loading, setLoading] = useState<boolean>(false)
+  const [error, setError] = useState<Error | null>(null)
+  const { lotusApi } = useEnvironment()
+
+  useEffect(() => {
+    setBalance(null)
+    setError(null)
+    if (!address) return
+    if (validateAddressString(address)) {
+      setLoading(true)
+      lotusApi
+        .request<string>('MsigGetAvailableBalance', address, null)
+        .then((b: string) => setBalance(new FilecoinNumber(b, 'attofil')))
+        .catch((e: Error) => setError(e))
+        .finally(() => setLoading(false))
+    } else {
+      setError(new Error('Invalid address'))
+    }
+  }, [address, lotusApi])
+
+  return { availableBalance: balance, loading, error }
+}
```
Separated this from useStateReadState

navFooh commented on src/utils/useStateReadState/index.ts:
```diff
@@ -1,123 +1,45 @@
-}
-
-/////// implemented until https://github.com/glifio/graph/issues/33
-export const useStateReadStateQuery = <T = any>(
-  baseOptions: QueryHookOptions<
-    T,
-    {
-      address: string
-    }
-  >
```
The name and params made it seem like this was an apollo query, which it's not. The params like `skip` that could be passed were not handled either, only the address was grabbed from the `baseOptions`.

navFooh commented on src/utils/useStateReadState/index.ts:
```diff
@@ -1,123 +1,45 @@
-      }
-    }
-
-    const firstLoad = !actorState && loading && !error
-    const newLoad = fetchedFor !== baseOptions?.variables?.address && !error
-
-    if (firstLoad || newLoad) {
```
This logic is removed now, but I wonder if we really need it. I can reimplement something similar if you think we do

navFooh commented on src/utils/useStateReadState/index.ts:
```diff
@@ -1,123 +1,45 @@
+  }, [address, lotusApi])
 
-  return { data: actorState, error, loading }
+  return { data, loading, notFound, error }
```
Return a `notFound` property instead of an empty object. Then we can show in the frontend what is going on instead of displaying an actor page with empty values. I was confused about this before.

navFooh committed:
Merge branch 'primary' into explorer-update

navFooh committed:
CollapsableLines

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update DataTypes.tsx

navFooh committed:
NullishLine

navFooh committed:
Update DataTypes.tsx

navFooh committed:
Update index.tsx

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
-  })
+    data: actorData,
+    loading: actorLoading,
+    notFound: actorNotFound,
+    error: actorError
+  } = useStateReadState(address)
 
+  // Load the address
   const {
     data: addressData,
-    error: addressError,
-    loading: addressLoading
+    loading: addressLoading,
+    error: addressError
   } = useAddressQuery({
-    variables: { address: convertAddrToPrefix(address, coinType) }
+    variables: { address }
   })
```
Not sure if we want to slow this down to prevent errors, but just to point out...

In some cases, a user might be viewing an SECP256K1 (`f1`) or BLS (`f3`) actor that does not yet exist on chain (no transactions have been sent from / to it). In this case, `StateReadState` would error 

However, thinking about it more, i think we should leave as-is - probably better to parallelize this and you've already handled the actor not found error

navFooh replied:
Yeah I imagine it will just show the actor not found. We can follow up with another PR if there is a better way to tackly this

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
 
-  const [viewActorState, setViewActorState] = useState(false)
+  // Get the actor name from the actor code
+  const actorName = useMemo<string | null>(
```
Do you think it would be useful to abstract the bulk of this logic out into its own `useActor` hook which could take an `address` and return the complete decorated actor state if it exists? It could also do some smarter (and aggregated) caching and polling. 

See my comment in the general review about redundant data fetching - if we stuck a `useActor` query into the react-components, the safe could just `useActor` too and share in the caching and revalidation benefits

The `proposal.tsx` would also benefit

navFooh replied:
Created an issue for this: https://github.com/glifio/react-components/issues/548

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
-          )}
-          <Line label='Actor'>{actorType}</Line>
+          <Line label='Actor name'>{actorName || 'unknown'}</Line>
+          <Line label='Actor code'>{actorData.Code['/']}</Line>
```
Can we `text-transform: capitalize` this?

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
+            <Line label='ID address'>{addressData?.address.id}</Line>
+          )}
           {addressData?.address.robust && (
             <Line label='Robust address'>{addressData?.address.robust}</Line>
```
Let's reverse these (show robust first)

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
+          <Line label='Actor code'>{actorData.Code['/']}</Line>
           <Line label='Balance'>
-            {new FilecoinNumber(actorStateData?.Balance, 'attofil').toFil()} FIL
+            {new FilecoinNumber(actorData.Balance, 'attofil').toFil()} FIL
```
Let's use `makePrettyBalance` here, 6 decimals

Schwartz10 commented on src/components/HistoryTables/ActorState/index.tsx:
```diff
@@ -1,118 +1,166 @@
+              ) : availableBalanceLoading ? (
+                <>Loading...</>
+              ) : availableBalance ? (
+                <>{availableBalance.toFil()} FIL</>
```
Let's use `makePrettyBalance` here, 6 decimals

Schwartz10 commented on src/components/LabeledText/AddressLink.tsx:
```diff
@@ -10,45 +10,43 @@ import {
+/**
+ * AddressLink is a LabeledLink for actor addresses.
+ *
+ * When fetchAddress is false (default), both id and address can be
```
2 Small nits:

1. Think the comment line here should be changed from: "but only the (robust) address will be shown if present" to "but the robust address takes precedent over the id address if they both exist"
2. i know this wasn't introduced in this PR, and it's probably inconsistent in other places too, but I feel like we should move towards naming consistency in:

```
address: {id: string, robust: string}
robust: string
id: string
```

The AddressLink uses `address` as a string, which can either be a robust or an ID, which feels inconsistent. 

Maybe we can add an optional `robust` param, and make the `address` type _either_ the GQL address type or a string, and handle both, so we can start using it correctly.

Happy to open a separate ticket for this thought if you agree

Schwartz10 commented on src/components/LabeledText/MessageLink.tsx:
```diff
@@ -9,16 +9,11 @@ import {
 } from './LabeledLink'
 
-export const CidLink = ({
+export const MessageLink = ({
```
Good call on this, was having similar thoughts before i saw this!

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+      <DataTypeLines
+        key={key}
+        label={key}
+        depth={label ? (depth ?? 0) + 1 : depth}
```
Think we could simplify things a bit if we give `depth` a `defaultProps` value of `1` - since (i think?), the only time you'd need to include a >1 `depth` value is recursively... the developer him/herself probably doesn't need to pass depth unless in a special case that we don't yet have?

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+  switch (dataType.Type) {
+    case Type.Bool:
+      const boolVal = Value as boolean
+      return <>{boolVal.toString()}</>
```
lol news to me... the boolean primitive has a toString method

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+        )
+
+      if (Name === 'FilecoinNumber')
+        return <>{new FilecoinNumber(strVal, 'attofil').toFil() + ' FIL'}</>
```
Let's `makeFriendlyBalance` with 6 decimals here?

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+
+      return <>{Value}</>
+
+    case Type.Number:
```
I think all Number types in the VM use 18 decimal precision, so they'll _all_ be in the `atto` SI prefix unit, meaning we should still be able to 

```
return <>{new FilecoinNumber(strVal, 'attofil').toFil()}
```

And just leave it without the "FIL" denominator at the end like the FilecoinNumber type has

navFooh replied:
I'm not sure if I understand correctly. When `dataType.Type === Type.Number`, it should just be a regular number, like a transaction ID or something, but not related to Filecoin values. Then it should always be a string.

Schwartz10 replied:
Ahh ok got it. I thought you were using Number for anything that wasn't specifically a Filecoin balance

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+      return <>{numVal}</>
+
+    case Type.Array:
+      const arrVal = Value as Array<BaseValue>
```
In our current use, can this ever actually happen? It looks like `DataTypeLines` maps over arrays and renders a `DataTypeLine` for each 

Either way, it seems like we should share similar behavior, and instead of joining array values by a string, we should just recurse through the array and render `DataTypeLines` for each array item like we do elsewhere?

Hard for me to get a good sense since i'm having trouble finding a case where this hits

navFooh replied:
That's true. I was considering a use case where we do want to display an array on a single line (maybe a bitfield), but that's not really relevant right now. I'll remove it and we can tackle it again if we need it

Schwartz10 commented on src/components/Layout/DataTypes.tsx:
```diff
@@ -0,0 +1,283 @@
+    case 'symbol':
+      return (
+        <Line label={label} depth={depth}>
+          {data.toString()}
```
What happens if we `toString` a function?

navFooh replied:
It should probably never occur that we have to display javascript functions here, but since this is like a fallback that should work in all cases, it doesn't hurt either.

```javascript
function test (foo, bar) {
    return foo + bar
}
test.toString() // function test (foo, bar) {\n    return foo + bar\n}
```

Schwartz10 commented on src/utils/useStateReadState/index.ts:
```diff
@@ -1,123 +1,45 @@
-  const [fetchedFor, setFetchedFor] = useState<string>('')
-
-  const apolloClient = useApolloClient()
+export const useStateReadState = <T = object | null>(
```
I think we should refactor to a `useSWR` approach here, like we should for `useMsigAvailablebalance` - comes equipped with more tools like polling, caching, and revalidation that will be useful for keeping actor states dynamically updating

Schwartz10 reviewed (changes requested):
1. The Proposal UI and the ActorState UI both use the "see more" dropdown links, but they look inconsistent:

<img width="949" alt="image" src="https://user-images.githubusercontent.com/12353734/188518680-318c3988-a2e0-472d-894c-34c1cce486ae.png">

<img width="687" alt="image" src="https://user-images.githubusercontent.com/12353734/188518691-7f5d3b23-f181-4b9e-83eb-9624594cdf3b.png">


Should we just take out the `<hr />`s that wrap the See More in the proposal detail?

2. Maybe better for a follow-on PR, but it seems like we do a lot of redundant data fetching in `Proposal.tsx` - this page currently _only_ exists inside the Safe, where the ActorName, ActorState...etc already exist. Seems like this could be solved nicely with a `useActor` hook that does a lot of the redundant work and caches + revalidates it
3. I think we should do a brief refactor of `useMsigGetAvailableBalance` and `useStateReadState` to a `useSWR` based approach - generally i've found it quite reliable, especially with built-in polling and revalidation functionality (which seems important for a dynamically refreshing /home UI in the safe). See example here https://github.com/glifio/react-components/blob/primary/src/services/UpdateBalance/index.tsx, makes it easy to write tests for this too
4. Testing - i think we should write some brief and simple tests for `useMsigAvailableBalance`, `useStateReadState`, and `ActorState` - hopefully the hooks testing should be quite simple because we have an example (thats probably more complicated) in `BalancePoller`, and we can mock most of it quite easily

navFooh commented:
> 1. The Proposal UI and the ActorState UI both use the "see more" dropdown links, but they look inconsistent:
My proposal would be to remove the "see more" toggle from the proposal altogether and just display the information that we're hiding by default. My understanding was that we want to limit the amount of space that the actor state consumes vertically so that we can still see the tx history below, but I don't feel this goes for the proposal. What it's hiding now is the next transaction ID and the list of people who already approved, which doesn't seem like too much info to me.

> 2. Maybe better for a follow-on PR, but it seems like we do a lot of redundant data fetching in `Proposal.tsx`
Let's do a follow up PR, I've created an issue: https://github.com/glifio/react-components/issues/548

> 3. I think we should do a brief refactor of `useMsigGetAvailableBalance` and `useStateReadState` to a `useSWR` based approach
Will have a look at this

> 4. Testing - i think we should write some brief and simple tests for `useMsigAvailableBalance`, `useStateReadState`, and `ActorState`
Will have a look at this

navFooh committed:
Remove Array handling in DataTypeValue

navFooh committed:
Use TypeScript for BalanceCard

navFooh committed:
update imports

navFooh committed:
linter fix

navFooh committed:
Remove pretty argument from makeFriendlyBalance

navFooh committed:
Make friendly balance update (#549)

navFooh committed:
robust first

navFooh committed:
makeFriendlyBalance

navFooh committed:
Capitalize the actor name for display

Schwartz10 reviewed (changes requested):
Hey this looks good @navFooh - two things:

1. Let's use an `useSWR` approach for `useMsigAvailableBalance` and `useStateReadState`
2. Let's add (just 1 is fine) test for `useMsigAvailableBalance` and `useStateReadState` (data should be mocked)
3. Let's add (probably 1 is fine as well) test for `ActorState`

Once you get those in, we can squash, merge, republish

navFooh committed:
useSWR and tests

export interface ActorStateProps

make actor state story tsx

Implement SWR for  useStateReadState

Use SWR for useMsigGetAvailableBalance

Add MULTISIG_ACTOR_ADDRESS to test utils

ensure loading is a boolean

useMsigGetAvailableBalance test

Update useMsigGetAvailableBalance test

Ensure boolean for notFound

useStateReadState test

Allow passing two string to isAddrEqual

Use isAddrEqual for tests

Add ActorState test

Linter fixes

navFooh committed:
Remove seeMore from Proposal
